veps/sig-compute/node-pressure-eviction.md

```diff
@@ -0,0 +1,229 @@
+
+There are ways to influence and even disable Node Pressure Eviction using the kubelet config, but be aware that these are node level configurations and not directly influenced by the workload itself. **Meaning, unlike API-Initiated Eviction, workload owners cannot directly influence how Node Pressure Eviction behaves for their workload.**
+
+## Motivation
+
+Cluster admins often have two initiatives that conflict with one another. On one hand they want to maximise VM workload density within their cluster to reduce cost. On the other hand they want to minimize disruption to those workloads. Having the ability to automate relieving node pressure by live migrating VMs helps with both of these initiatives.
+
+## Goals
+Currently, the VMI api’s EvictionStrategy tunable only works with API-Initiated Eviction. The goal is to extend the behavior of the EvictionStrategy to Node Pressure Eviction as well. This means if a VMI has an EvictionStrategy of LiveMigrate, we expect the system to attempt to live migrate the VMI regardless of which eviction method initiated the eviction.
+
+## Non Goals
+* There is no expectation that NodePressure and API-Initiated eviction will provide the same guarantees.
+
+## Repos
+
+- kubevirt/kubevirt - this is where the majority of the node pressure eviction logic will be implemented
+- kubevirt/hyperconverged-operator - this component will need to expose enabling the feature gate for node pressure eviction handling
+
+## Definition of Users
+- Cluster Admins who wish to influence at the cluster level how VMs behave during Node Pressure Eviction
+
+## User Stories
+* As a cluster admin, I would prefer that VMs across the cluster default to attempting best effort live migrated during Node Pressure Eviction rather than being restarted.
+
+# Design
+This section outlines the theory behind how this design works and the code changes required to implement this theory into practice.
+
+## Design Theory
+
+The theory behind this design can be summed up in the following points.
+
+* Currently a user expresses their desire to perform an API initiated shutdown of a VMI by deleting that VMI. This results in vmi.DeletionTimestamp != nil.
+* Currently when the virt-launcher pod intercepts a signal to shutdown (SIG_TERM), it passes that knowledge to virt-handler and virt-handler coordinates how to shutdown the VMI.
+* This means if virt-handler is told by virt-launcher that a shutdown signal has been received, and the vmi.DeletionTimestamp == nil, then this shutdown is not an expression of the user’s intent but instead an expression of the systems attempt to evict the workload.
+
+Essentially, if the system observes a virt-launcher receive a shutdown signal (like SIG_TERM), but the vmi.DeletionTimestamp == nil, then the system can process that as an eviction.
```

Comment:
Does this mean an evacuation migration is triggered if a user deletes the virt-launcher pod?

Reply:
as the design is written now, yes.

If this is a situation we want to avoid (not convinced either way yet), one solution might be that the evacuation controller makes the decision to shutdown the VMI instead of live migrating it if the pod's deletionTimestamp != nil.

Reply:
I still don't know if it's something we want to avoid or not :slightly_smiling_face:. I just wanted to make sure of that, thanks. My only concern is that a user can trigger a migration with a simple pod deletion, and this can affect other workloads. Maybe we can tweak things better with priority queues.

Reply:
> My only concern is that a user can trigger a migration with a simple pod deletion

yep. it's a good observation. I believe we have similar indirect ways of triggering live migration with the live update feature today. It's possible this use case is a bit different because we'd want ensure these migrations have the highest priority due to them being time boxed.

I think it's reasonable to have the evacuation controller tear down the VMI instead of live migrating it if the pod's deletion timestamp is set but the VMI's is not.